src/treelearner/feature_histogram.hpp

```diff
@@ -211,13 +213,28 @@ class FeatureHistogram {
 
 private:
   /*!
+  * \brief Return the L1 regularized sum_gradients
+  * \param sum_gradients
+  * \return regularized sum_gradients
+  */
+  double L1RegularizedGradients(double sum_gradients) const {
+  if (sum_gradients > lambda_l1_) return sum_gradients - lambda_l1_;
+  if (sum_gradients < -lambda_l1_) return sum_gradients + lambda_l1_;
+  return 0.0f;
+  }
+
+  /*!
   * \brief Calculate the split gain based on sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return split gain
   */
   double GetLeafSplitGain(double sum_gradients, double sum_hessians) const {
-    return (sum_gradients * sum_gradients) / (sum_hessians);
+    if (lambda_l1_ != 0.0f) {
```

Comment:
I'll update pull with the changes proposed here. This'll also make it possible to immediately return `0.0f` from `GetLeafSplitGain()` and `CalculateSplittedLeafOutput()` when `!((sum_gradients > lambda_l1_) || (sum_gradients < lambda_l1_))`, thereby sidestepping quite a bit of computation.

As for reducing the number of `if`'s, we could reduce the calculations to:
```c++
if (abs(sum_gradients) > lambda_l1_) {
  return sign(sum_gradients) * (abs(sum_gradients) - lambda_l1_));
}
return 0.0f
```
A special version for `GetLeafSplitGain()` could then be a bit faster by omitting the multiplication with `sign(sum_gradients)`, because the output is squared afterwards anyway.

Nevertheless, it really comes down to how fast we can obtain the sign of `sum_gradients` in other cases and how fast a call to `std::abs` really is. There are some bit-shifts we can use for the sign, or it could be that the compiler knows a faster way to do `(x > 0) - (x < 0)`, but I will have to do some profiling to see if this is faster than the `if`-statements we have now.